apiserver/common/life.go

```diff
@@ -11,27 +11,72 @@ import (
 
 	apiservererrors "github.com/juju/juju/apiserver/errors"
 	"github.com/juju/juju/core/life"
+	coreunit "github.com/juju/juju/core/unit"
+	applicationerrors "github.com/juju/juju/domain/application/errors"
 	"github.com/juju/juju/rpc/params"
 	"github.com/juju/juju/state"
 )
 
 // LifeGetter implements a common Life method for use by various facades.
 type LifeGetter struct {
-	st         state.EntityFinder
-	getCanRead GetAuthFunc
+	st                 state.EntityFinder
+	applicationService LifeGetterApplicationService
+	getCanRead         GetAuthFunc
+}
+
+// LifeGetterApplicationService provides application domain service methods for
+// getting the life of applications and units.
+type LifeGetterApplicationService interface {
+	// GetUnitLife looks up the life of the specified unit, returning an error
+	// satisfying [applicationerrors.UnitNotFoundError] if the unit is not found.
+	GetUnitLife(ctx context.Context, unitName coreunit.Name) (life.Value, error)
+	// GetApplicationLifeByName looks up the life of the specified application, returning
+	// an error satisfying [applicationerrors.ApplicationNotFoundError] if the
+	// application is not found.
+	GetApplicationLifeByName(ctx context.Context, appName string) (life.Value, error)
 }
 
 // NewLifeGetter returns a new LifeGetter. The GetAuthFunc will be used on
 // each invocation of Life to determine current permissions.
-func NewLifeGetter(st state.EntityFinder, getCanRead GetAuthFunc) *LifeGetter {
+func NewLifeGetter(
+	st state.EntityFinder,
+	getCanRead GetAuthFunc,
+	applicationService LifeGetterApplicationService,
```

Comment:
This is just a thought: since we will usually be instantiating these life getters when instantiating the different APIs, could this receive a `domainServices` param instead of each needed service? (I understand this will take more and more services right?)

Reply:
Simon's PR has pulled in more. Next person here can do this.